### PR TITLE
An openshift template that builds apicast from source

### PR DESCRIPTION
**Pull request by nmasse-itix**
When working on Apicast, such as developing policies or customizations, it might be good to have Continuous Integration and sometimes Continuous Delivery. 

This pull requests is a template for OpenShift that enables CI/CD. Once the template instanciated, a two-stage buildconfig is created (first a custom builder, then the apicast build using this custom builder). 

If you want to give it a try, on an OCP cluster you can use : 
```
oc process -f build-template.yaml -p GIT_REPO=https://github.com/nmasse-itix/3scale-apicast.git -p GIT_REFERENCE=openshift-builder-latest -p OPENSHIFT_BASE_IMAGE=registry.access.redhat.com/openshift3/ose:v3.7 |oc create -f -
```

Or on a minishift/OpenShift origin, you can use :
```
oc process -f build-template.yaml -p GIT_REPO=https://github.com/nmasse-itix/3scale-apicast.git -p GIT_REFERENCE=openshift-builder-latest |oc create -f -
```

The build can then be triggered from Jenkins, as usual.

### CODE REVIEW

**Comment by mikz** [comment]
Why using a custom builder instead of using our s2i-openresty images?

```yaml
apiVersion: v1
kind: Template
metadata:
  name: "apicast-build"
message: "APIcast BuildConfig Template"
objects:

- apiVersion: v1
  kind: BuildConfig
  metadata:
    labels:
      app: apicast
    name: apicast
  spec:
    output:
      to:
        kind: ImageStreamTag
        name: amp-apicast:latest
    source:
      contextDir: gateway
      git:
        uri: https://github.com/3scale/apicast.git
        ref: "${GIT_REF}"
      type: Git
    strategy:
      sourceStrategy:
        forcePull: true
        from:
          kind: ImageStreamTag
          name: s2i-openresty-centos7:builder
        runtimeImage:
          kind: ImageStreamTag
          name: s2i-openresty-centos7:runtime
      type: Source
    triggers:
    - type: ConfigChange
    - type: ImageChange

- apiVersion: v1
  kind: ImageStream
  metadata:
    labels:
      app: apicast
    name: s2i-openresty-centos7
  spec:
    tags:
    - from:
        kind: DockerImage
        name: quay.io/3scale/s2i-openresty-centos7:1.13.6.1-rover5
      name: builder
    - from:
        kind: DockerImage
        name: quay.io/3scale/s2i-openresty-centos7:1.13.6.1-rover5-runtime
      name: runtime


parameters:
- name: GIT_REF
  description: "Git Reference to use. Can be a tag or branch."
  value: "master"
  required: true
```

**Comment by nmasse-itix** [comment]
Hummmm, where did you get this parameter "runtimeImage" ? :D 
It does not appear anywhere in the documentation...

**Comment by mikz** [comment]
@nmasse-itix checking the rest API object docs: https://docs.openshift.com/container-platform/3.6/rest_api/openshift_v1.html#v1-sourcebuildstrategy
Looks like that is gone in 3.7. That object spec doc was basically my reference for everything.

**Comment by nmasse-itix** [comment]
OK, thanks for the tip ! 👍 

I used the custom builder only as a bootstrap to call the S2I scripts using the required parameters (runtimeImage for instance). 

Since it seems there is a proper way to do it directly with a S2I build, I suggest to close this pull request and open another one with just your template. 

I think it is important to get it in the official apicast repo to help people save time since the build config is not trivial. What do you think ?

**Comment by mikz** [comment]
@nmasse-itix I agree it is something that should be provided.
Unfortunately we are a bit busy right now  polishing the next release. 

Would be awesome If you'd have the time to take the example I sent, try it out and just force push to this branch.
Maybe something is missing and I'm not really sure what should be the best practice regarding the output (custom registry or image tag) and how to maintain the OpenResty version in sync with our CI pipeline.
But something should be better than nothing!

**Comment by nmasse-itix** [comment]
I tested your template. It work like a charm, thanks a lot for your help and suggestion ! :)
I force-pushed the new template. 

Does it seems acceptable for a merge ?

File: openshift/apicast-build-template.yaml [hunk 1]
```diff
@@ -0,0 +1,87 @@
+apiVersion: v1
+kind: Template
+labels:
+  template: apicast-build
+metadata:
+  annotations:
+    description: |-
+      A BuildConfig for OpenShift that builds images of Apicast from source.
```
Comment (mikz) [inline openshift/apicast-build-template.yaml]: @maneta do we want to do something with those annotations? Adding some metadata, tags, ...

File: openshift/apicast-build-template.yaml [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    openshift.io/display-name: Apicast BuildConfig
+    template.openshift.io/documentation-url: https://github.com/3scale/apicast
+    template.openshift.io/long-description: Builds images of Apicast
+    template.openshift.io/provider-display-name: 3scale
+    template.openshift.io/support-url: https://github.com/3scale/apicast/issues
+  name: apicast-build
+parameters:
+- description: The Apicast GIT repository to build.
```
Comment (mikz) [inline openshift/apicast-build-template.yaml]: A tiny nitpick. We are trying to spell it as `APIcast`.

File: openshift/apicast-build-template.yaml [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+- description: The GIT reference (branch/tag) to use.
+  displayName: GIT Reference
+  name: GIT_REF
+  value: master
+
+- description: The OpenResty S2I Builder Image to use.
+  displayName: OpenResty S2I Builder Image
+  name: OPENRESTY_BUILDER_IMAGE
+  value: quay.io/3scale/s2i-openresty-centos7:master
```
Comment (mikz) [inline openshift/apicast-build-template.yaml]: Good idea 👍

**Comment by mikz** [comment]
@nmasse-itix looks pretty good to me. Just the APIcast spelling. And I'll need review from @maneta because he is now working on updating templates for OpenShift Online.

File: openshift/apicast-build-template.yaml (hunk 1 above)
Comment (maneta) [reply openshift/apicast-build-template.yaml]: yes, the guidelines I have from product are:

```yaml
openshift.io/display-name: 3scale APIcast API Gateway
openshift.io/provider-display-name: Red Hat, Inc.
iconClass: icon-3scale
description: >-
  3scale's APIcast is an NGINX based API gateway used to integrate your internal and external
  API services with 3scale's API Management Platform. It supports OpenID connect to integrate
  with external Identity Providers such as Red Hat Single Sign On, for API traffic authentication.
```

File: openshift/apicast-build-template.yaml [hunk 4]
```diff
@@ -0,0 +1,87 @@
+apiVersion: v1
+kind: Template
+labels:
+  template: apicast-build
+metadata:
+  annotations:
```
Comment (maneta) [inline openshift/apicast-build-template.yaml]: Maybe should update those annotations with the product guidelines:

```yaml
openshift.io/display-name: 3scale APIcast API Gateway
openshift.io/provider-display-name: Red Hat, Inc.
iconClass: icon-3scale
description: >-
  3scale's APIcast is an NGINX based API gateway used to integrate your internal and external
  API services with 3scale's API Management Platform. It supports OpenID connect to integrate
  with external Identity Providers such as Red Hat Single Sign On, for API traffic authentication.
```

File: openshift/apicast-build-template.yaml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+objects:
+
+- apiVersion: v1
+  kind: ImageStream
+  metadata:
+    labels:
+      app: apicast
+    name: apicast
+  spec:
```
Comment (maneta) [inline openshift/apicast-build-template.yaml]: Here we could just define the latest tag like this:

```yaml
spec:
  tags:
  - name: latest
    annotations:
      openshift.io/display-name: APIcast (latest) 
    from:
      kind: ImageStreamTag
      name: latest
```

File: openshift/apicast-build-template.yaml [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        uri: "${GIT_REPO}"
+        ref: "${GIT_REF}"
+      type: Git
+    strategy:
+      sourceStrategy:
+        forcePull: true
+        from:
+          kind: ImageStreamTag
+          name: s2i-openresty-centos7:builder
+        runtimeImage:
```
Comment (mikz) [inline openshift/apicast-build-template.yaml]: Looks like this is gone in 3.7.1:

> "found invalid field runtimeImage for v1.SourceBuildStrategy" 

Sigh. The documentation says to use ImageSource.

**Comment by eloycoto** [comment]
Already covered by AMP templates.

**Comment by maneta** [comment]
LGTM!

**Pull request closed without merging**